src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_bignum.c

```diff
@@ -20,12 +20,41 @@ int32_t CryptoNative_BigNumToBinary(jobject bignum, uint8_t* output)
     JNIEnv* env = GetJNIEnv();
     jbyteArray bytes = (jbyteArray)(*env)->CallObjectMethod(env, bignum, g_toByteArrayMethod);
     jsize bytesLen = (*env)->GetArrayLength(env, bytes);
-    (*env)->GetByteArrayRegion(env, bytes, 0, bytesLen, (jbyte*)output);
+
+    // We strip the leading zero byte from the byte array.
+    jsize startingIndex = 0;
+    jbyte leadingByte;
+    (*env)->GetByteArrayRegion(env, bytes, 0, 1, &leadingByte);
+    if (leadingByte == 0)
+    {
+        startingIndex++;
+        bytesLen--;
+    }
+
+    (*env)->GetByteArrayRegion(env, bytes, startingIndex, bytesLen, (jbyte*)output);
     (*env)->DeleteLocalRef(env, bytes);
     return CheckJNIExceptions(env) ? FAIL : (int32_t)bytesLen;
 }
 
 int32_t CryptoNative_GetBigNumBytes(jobject bignum)
+{
+    // bigNum.toByteArray().length();
+    JNIEnv* env = GetJNIEnv();
+    jbyteArray bytes = (jbyteArray)(*env)->CallObjectMethod(env, bignum, g_toByteArrayMethod);
+    jsize bytesLen = (*env)->GetArrayLength(env, bytes);
+
+    // We strip the leading zero byte from the byte array.
+    jbyte leadingByte;
+    (*env)->GetByteArrayRegion(env, bytes, 0, 1, &leadingByte);
```

Comment:
Possibly cheaper to use the `signum()` function, and subtract one if it returns a negative number?

Reply:
https://docs.oracle.com/javase/7/docs/api/java/math/BigInteger.html#signum()

Reply:
The issue here is that the BigInteger is positive due to the leading zero byte, so the signum function here isn't helpful.

Reply:
Ah, right, it's the positive numbers where the high bit in the high byte is set.  Not sure what negative numbers I was thinking of.

Still possibly more performant than the toByteArray method is using bitLength (if it's a multiple of 8 and signum is >0 then you'll get the padding byte). At least, assuming I read the docs correctly :smile:.

Reply:
I'll test out some cases and see if I can do this more efficiently.

Reply:
I tested it out and this won't work. `bitlength` doesn't count the sign bit even when the BigInteger is `0xffffff` and forced positive by the sign+magnitude constructor.

I think we have to do it the way I'm doing it now.

Reply:
> bitlength doesn't count the sign bit

Yeah, that matches the docs: "Returns the number of bits in the minimal two's-complement representation of this BigInteger, excluding a sign bit.".

I think it's something like

```
int bitLen = bigInt.bitLength();
return (bitLen + 7) / 8;
```

(Might be wrong for negative numbers if it really "excludes" the sign bit, vs excludes the cleared sign bit for positive numbers... then `if (bigInt.signum() < 0) { bitLen++; }`)

My expectation from the docs is that positive 0xFF will report signum() as 1 and bitLength() as 8, but export in two bytes.

(8 + 7) / 8 == 15 / 8 == 1

The extractor wouldn't need to change.  This just detects the same thing without going through array export.

Reply:
Yes, your expectation is correct.

I'll update this to use the `(bitlength+7) / 8` calculation for the case that strips the leading zero byte.